src/base/utils/sys.cpp

```diff
@@ -0,0 +1,218 @@
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "sys.h"
+
+#ifdef Q_OS_WIN
+#include <ShlObj.h>
+#include <shobjidl.h>
+
+#include <QCoreApplication>
+#include <QMessageBox>
+#include <QSettings>
+#include <QSysInfo>
+
+#include "base/preferences.h"
+
+namespace
+{
+    LPCWSTR APP_REGISTRY_NAME = L"qBittorrent";
+
+    inline QString tr(const char *message)
+    {
+        return qApp->translate("System", message);
+    }
+
+    void showInitialAssocUIImpl1()
+    {
+        using namespace Utils::Sys;
+
+        const QString dlgTitle {tr("Torrent file association")};
+        const QString dlgText {
+            tr("qBittorrent is not the default application to open torrent files or Magnet links.\n"
+               "Do you want to associate qBittorrent to torrent files and Magnet links?")
+        };
+
+        if (!isTorrentFileAssocSet() || !isMagnetLinkAssocSet())
+        {
+            if (QMessageBox::question(nullptr, dlgTitle, dlgText, QMessageBox::Yes | QMessageBox::No, QMessageBox::Yes) == QMessageBox::Yes) {
+                setTorrentFileAssoc();
+                setMagnetLinkAssoc();
+            }
+            else {
+                Preferences::instance()->setNeverCheckFileAssoc();
+            }
+        }
+    }
+
+#ifdef APP_ASSOC_REG
+    void showInitialAssocUIImpl2()
+    {
+        const QString dlgTitle {tr("Torrent file association")};
+        const QString dlgText {
+            tr("qBittorrent is not the default application to open torrent files or Magnet links.\n"
+               "Do you want to associate qBittorrent to torrent files and Magnet links?")
+        };
+
+        HRESULT hr = ::CoInitializeEx(NULL, COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE);
+        if (FAILED(hr)) return;
+
+        IApplicationAssociationRegistration *pAssoc;
+
+        hr = ::CoCreateInstance(
+                 CLSID_ApplicationAssociationRegistration, 0, CLSCTX_INPROC
+                 ,__uuidof(IApplicationAssociationRegistration), reinterpret_cast<void**>(&pAssoc));
+        if (SUCCEEDED(hr)) {
+            BOOL isDefault = TRUE;
+
+            hr = pAssoc->QueryAppIsDefaultAll(AL_EFFECTIVE, APP_REGISTRY_NAME, &isDefault);
+            if ((SUCCEEDED(hr) && (isDefault == FALSE)) || (hr == HRESULT_FROM_WIN32(ERROR_NO_ASSOCIATION))) {
+                if (QMessageBox::question(nullptr, dlgTitle, dlgText, QMessageBox::Yes | QMessageBox::No, QMessageBox::Yes) == QMessageBox::Yes)
+                    pAssoc->SetAppAsDefaultAll(APP_REGISTRY_NAME);
+                else
+                    Preferences::instance()->setNeverCheckFileAssoc();
+            }
+
+            pAssoc->Release();
+        }
+
+        ::CoUninitialize();
+    }
+
+    void showInitialAssocUIImpl3()
```

Comment:
1, 2, 3? Are you kidding?

Reply:
Yes. (What a question, that's the answer!)